c2rust-analyze/src/main.rs

```diff
@@ -474,6 +474,24 @@ fn run(tcx: TyCtxt) {
                 || info.contains(PointerInfo::NOT_TEMPORARY_REF))
     }
 
+    // track all types mentioned in extern blocks, we
+    // don't want to rewrite those
+    let mut foreign_mentioned_tys = HashSet::new();
+    for item in tcx.hir_crate_items(()).foreign_items() {
+        let ldid = item.def_id.to_def_id();
+        match tcx.def_kind(ldid) {
+            DefKind::Fn | DefKind::AssocFn => {
+                let sig = tcx.fn_sig(ldid);
+                let sig = tcx.erase_late_bound_regions(sig);
+                for ty in sig.inputs_and_output {
+                    foreign_mentioned_tys.insert(ty);
+                }
```

Comment:
After reading @kkysen's comment https://github.com/immunant/c2rust/pull/960#discussion_r1229986055, I think this should actually recurse into fields as well (for `#[repr(C)]` structs at minimum).
```Rust
#[repr(C)]
struct Foo {
    x: *mut i32,
    y: *mut i32,
}
#[repr(C)]
struct Bar {
    foo: *mut Foo,
}
extern "C" {
    fn f(bar: *mut Bar);
}
```
Here, it's possible the C code for `f` might access `bar->foo->y`, in which case it relies on the layout of `Foo`, even though `Foo` itself isn't mentioned in any `extern` signatures.  So all the fields of `Foo` would need to be FIXED in this case.

Reply:
Does `rustc` have a `fn` to do this already?  As `rustc` checks (as a warning) to make sure your FFI-crossing types are FFI-safe.

Reply:
I'm not sure.  It doesn't look like `TypeFolder` will walk into fields, at least.  It might be worth looking at the code for the FFI-safety lint to see if it's built on some kind of reusable framework.

Reply:
> Does `rustc` have a `fn` to do this already? As `rustc` checks (as a warning) to make sure your FFI-crossing types are FFI-safe.

It doesn't seem so, at least not in a way neatly separated from the lint.

Reply:
Walking in 795438ae